qiskit_experiments/framework/experiment_data.py

```diff
@@ -1070,7 +1070,11 @@ def _add_result_data(self, result: Result, job_id: Optional[str] = None) -> None
     def _retrieve_data(self):
         """Retrieve job data if missing experiment data."""
         # Get job results if missing in experiment data.
-        if self.provider is None:
+        if self.provider is None and not self._result_data.copy():
+            LOG.warning(
+                "provider is None and there is no result data that are stored."
+                " no data was retrieved."
+            )
```

Comment:
1. Because `_result_data` can be locked so I used the copy method.
2. I used `and not self._result_data.copy()` so it will not print the warning unnecessarily. If there is no provider but there are results in ` self._result_data`, the warning is meaningless.
3. What about `Provider for ExperimentData object doesn't exist. Failed to retrieve data from the server.` ?

Reply:
1. `_result_data` is an attribute of `ExperimentData` of type `ThreadSafeList()` and is used as a list of `Result` object. As a thread safe object, whenever `_result_data` is being is being accessed (append data, read data, etc) it is needed to be locked (Source for working with safe thread obj [here](https://superfastpython.com/thread-safe-list/)). The `copy` method is taking care of it automatically so I used it for clear code reading.
2. True, I changed it to 
```
  if self.provider is None:
      if not self._result_data.copy():
          # Adding warning so the user will have indication why the analysis may fail.
          LOG.warning("Warning Text.")
      return
```
3. I am not sure if to include it in the warning as this warning is there to indicate the user that the problem in analysis could be due to missing data. What about the following sentence? "Provider for ExperimentData object doesn't exist, resulting in a failed attempt to retrieve data from the server ; no stored result data exists"

Reply:
(3) is very good. Only remove the space between "server" and the semicolon that follows it.

(2) is also good.

I still don't understand (1). Is `_result_data` already locked, and for some reason you want to overcome it, and the copying helps you with this? Or does `copy` lock, in this case, why do you want to lock? And when will it be unlocked? And what's locked - the original or the copy? Anyhow this should be explained in a code comment.

Reply:
I understand. The original `_result_data` is locked, therefore the check `if _result_data` is impossible. So you copy it, the new object is not locked and you can check `if _result_data.copy()`. So just please an inline comment in the code.